File: server/mcp-server.js
#!/usr/bin/env node

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";
import pg from 'pg';

const { Pool } = pg;

const connectionString = "postgresql://finflow_db_ggfz_user:4KUb9md97Vcufc1eRTP4VIRL77FaP4r4@dpg-d6sgr3juibrs73eb9hig-a.singapore-postgres.render.com/finflow_db_ggfz?ssl=true";

const pool = new Pool({
  connectionString,
  ssl: { rejectUnauthorized: false }
});

pool.on('error', (err) => {
  console.error('Unexpected error on idle client:', err.message);
});

const CURRENT_USER_ID = "cmnlkvkp30000hqm96cwu47rc";

const server = new Server({
  name: "finflow-backend-agent",
  version: "2.0.0",
}, {
  capabilities: { tools: {} },
});

// ─────────────────────────────────────────────
// TOOLS LIST
// ─────────────────────────────────────────────
server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [

    // ── BALANCE ──
    {
      name: "get_balance",
      description: "Get the current total balance (Income - Expenses)",
      inputSchema: { type: "object", properties: {} }
    },

    // ── TRANSACTIONS ──
    {
      name: "get_transactions",
      description: "Get transactions with optional filters",
      inputSchema: {
        type: "object",
        properties: {
          type: { type: "string", enum: ["income", "expense"] },
          category: { type: "string" },
          from_date: { type: "string", description: "ISO date string" },
          to_date: { type: "string", description: "ISO date string" },
          limit: { type: "number", description: "Default 50" }
        }
      }
    },
    {
      name: "add_income",
      description: "Log a new income entry",
      inputSchema: {
        type: "object",
        properties: {
          amount: { type: "number" },
          description: { type: "string" },
          category: { type: "string" },
          budget_id: { type: "string", description: "Optional budget to link" }
        },
        required: ["amount", "description", "category"]
      }
    },
    {
      name: "add_expense",
      description: "Log a new expense entry",
      inputSchema: {
        type: "object",
        properties: {
          amount: { type: "number" },
          description: { type: "string" },
          category: { type: "string" },
          budget_id: { type: "string", description: "Optional budget to link and auto-update spent" }
        },
        required: ["amount", "description", "category"]
      }
    },
    {
      name: "update_transaction",
      description: "Update an existing transaction",
      inputSchema: {
        type: "object",
        properties: {
          id: { type: "string" },
          amount: { type: "number" },
          description: { type: "string" },
          category: { type: "string" },
          type: { type: "string", enum: ["income", "expense"] }
        },
        required: ["id"]
      }
    },
    {
      name: "delete_transaction",
      description: "Delete a transaction by ID",
      inputSchema: {
        type: "object",
        properties: { id: { type: "string" } },
        required: ["id"]
      }
    },
    {
      name: "get_spending_summary",
      description: "Get spending grouped by category for a given month/year",
      inputSchema: {
        type: "object",
        properties: {
          month: { type: "number" },
          year: { type: "number" }
        }
      }
    },
    {
      name: "get_monthly_trend",
      description: "Get income vs expense totals month by month",
      inputSchema: {
        type: "object",
        properties: {
          months: { type: "number", description: "Past N months, default 6" }
        }
      }
    },

    // ── BUDGETS ──
    {
      name: "get_budgets",
      description: "Get all budgets, optionally filtered by month/year",
      inputSchema: {
        type: "object",
        properties: {
          month: { type: "number" },
          year: { type: "number" }
        }
      }
    },
    {
      name: "create_budget",
      description: "Create a new budget for a category and month",
      inputSchema: {
        type: "object",
        properties: {
          name: { type: "string" },
          amount: { type: "number" },
          category: { type: "string" },
          month: { type: "number" },
          year: { type: "number" }
        },
        required: ["name", "amount", "category", "month", "year"]
      }
    },
    {
      name: "update_budget",
      description: "Update an existing budget's name, amount, or category",
      inputSchema: {
        type: "object",
        properties: {
          id: { type: "string" },
          name: { type: "string" },
          amount: { type: "number" },
          category: { type: "string" }
        },
        required: ["id"]
      }
    },
    {
      name: "delete_budget",
      description: "Delete a budget by ID",
      inputSchema: {
        type: "object",
        properties: { id: { type: "string" } },
        required: ["id"]
      }
    },

    // ── LOANS ──
    {
      name: "get_loans",
      description: "Get all loans for the current user",
      inputSchema: { type: "object", properties: {} }
    },
    {
      name: "add_loan",
      description: "Add a new loan",
      inputSchema: {
        type: "object",
        properties: {
          name: { type: "string" },
          type: { type: "string", enum: ["home", "car", "personal", "education", "other"] },
          principal: { type: "number" },
          outstanding: { type: "number" },
          emi: { type: "number" },
          interest_rate: { type: "number" },
          tenure_months: { type: "number" },
          start_date: { type: "string", description: "ISO date string" }
        },
        required: ["name", "type", "principal", "outstanding", "emi", "interest_rate", "tenure_months", "start_date"]
      }
    },
    {
      name: "update_loan",
      description: "Update an existing loan (e.g. after EMI payment, update outstanding)",
      inputSchema: {
        type: "object",
        properties: {
          id: { type: "string" },
          outstanding: { type: "number" },
          emi: { type: "number" },
          interest_rate: { type: "number" }
        },
        required: ["id"]
      }
    },
    {
      name: "delete_loan",
      description: "Delete a loan by ID",
      inputSchema: {
        type: "object",
        properties: { id: { type: "string" } },
        required: ["id"]
      }
    }

  ]
}));

// ─────────────────────────────────────────────
// TOOL HANDLERS
// ─────────────────────────────────────────────
server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args = {} } = request.params;

  try {

    // ── BALANCE ──
    if (name === "get_balance") {
      const result = await pool.query(
        `SELECT
           SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END) as income,
           SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END) as expenses
         FROM "Transaction" WHERE "userId" = $1`,
        [CURRENT_USER_ID]
      );
      const { income = 0, expenses = 0 } = result.rows[0];
      const balance = (Number(income) - Number(expenses)).toFixed(2);
      return { content: [{ type: "text", text: `Balance: ₹${balance} (Income: ₹${Number(income).toFixed(2)}, Expenses: ₹${Number(expenses).toFixed(2)})` }] };
    }

    // ── GET TRANSACTIONS ──
    if (name === "get_transactions") {
      const { type, category, from_date, to_date, limit = 50 } = args;
      let query = `SELECT * FROM "Transaction" WHERE "userId" = $1`;
      const params = [CURRENT_USER_ID];
      let i = 2;
      if (type)      { query += ` AND type = $${i++}`;           params.push(type); }
      if (category)  { query += ` AND category = $${i++}`;       params.push(category); }
      if (from_date) { query += ` AND date >= $${i++}`;          params.push(from_date); }
      if (to_date)   { query += ` AND date <= $${i++}`;          params.push(to_date); }
      query += ` ORDER BY date DESC LIMIT $${i}`;                params.push(limit);
      const result = await pool.query(query, params);
      return { content: [{ type: "text", text: JSON.stringify(result.rows, null, 2) }] };
    }

    // ── ADD INCOME ──
    if (name === "add_income") {
      const { amount, description, category, budget_id } = args;
      await pool.query(
        `INSERT INTO "Transaction" (id, description, amount, type, category, date, source, "userId", "budgetId")
         VALUES (gen_random_uuid(), $1, $2, 'income', $3, NOW(), 'ai_parsed', $4, $5)`,
        [description, amount, category, CURRENT_USER_ID, budget_id || null]
      );
      return { content: [{ type: "text", text: `✅ Income of ₹${amount} logged: "${description}" under ${category}.` }] };
    }

    // ── ADD EXPENSE ──
    if (name === "add_expense") {
      const { amount, description, category, budget_id } = args;
      await pool.query(
        `INSERT INTO "Transaction" (id, description, amount, type, category, date, source, "userId", "budgetId")
         VALUES (gen_random_uuid(), $1, $2, 'expense', $3, NOW(), 'ai_parsed', $4, $5)`,
        [description, amount, category, CURRENT_USER_ID, budget_id || null]
      );
      if (budget_id) {
        await pool.query(
          `UPDATE "Budget" SET spent = spent + $1, "updatedAt" = NOW() WHERE id = $2 AND "userId" = $3`,
          [amount, budget_id, CURRENT_USER_ID]
        );
      }
      return { content: [{ type: "text", text: `✅ Expense of ₹${amount} logged: "${description}" under ${category}.` }] };
    }

    // ── UPDATE TRANSACTION ──
    if (name === "update_transaction") {
      const { id, amount, description, category, type } = args;
      const updates = [];
      const params = [];
      let i = 1;
      if (amount !== undefined) { updates.push(`amount = $${i++}`);      params.push(amount); }
      if (description)          { updates.push(`description = $${i++}`); params.push(description); }
      if (category)             { updates.push(`category = $${i++}`);    params.push(category); }
      if (type)                 { updates.push(`type = $${i++}`);        params.push(type); }
      if (updates.length === 0) return { content: [{ type: "text", text: "No fields to update." }] };
      params.push(id, CURRENT_USER_ID);
      await pool.query(
        `UPDATE "Transaction" SET ${updates.join(", ")} WHERE id = $${i++} AND "userId" = $${i}`,
        params
      );
      return { content: [{ type: "text", text: `✅ Transaction ${id} updated.` }] };
    }

    // ── DELETE TRANSACTION ──
    if (name === "delete_transaction") {
      await pool.query(`DELETE FROM "Transaction" WHERE id = $1 AND "userId" = $2`, [args.id, CURRENT_USER_ID]);
      return { content: [{ type: "text", text: `🗑️ Transaction ${args.id} deleted.` }] };
    }

    // ── SPENDING SUMMARY ──
    if (name === "get_spending_summary") {
      const { month, year } = args;
      let query = `SELECT category, SUM(amount) as total FROM "Transaction"
                   WHERE "userId" = $1 AND type = 'expense'`;
      const params = [CURRENT_USER_ID];
      let i = 2;
      if (month) { query += ` AND EXTRACT(MONTH FROM date) = $${i++}`; params.push(month); }
      if (year)  { query += ` AND EXTRACT(YEAR FROM date) = $${i++}`;  params.push(year); }
      query += ` GROUP BY category ORDER BY total DESC`;
      const result = await pool.query(query, params);
      return { content: [{ type: "text", text: JSON.stringify(result.rows, null, 2) }] };
    }

    // ── MONTHLY TREND ──
    if (name === "get_monthly_trend") {
      const { months = 6 } = args;
      const result = await pool.query(
        `SELECT TO_CHAR(date, 'YYYY-MM') as month,
                SUM(CASE WHEN type = 'income'  THEN amount ELSE 0 END) as income,
                SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END) as expenses
         FROM "Transaction"
         WHERE "userId" = $1 AND date >= NOW() - ($2 || ' months')::INTERVAL
         GROUP BY TO_CHAR(date, 'YYYY-MM')
         ORDER BY month ASC`,
        [CURRENT_USER_ID, months]
      );
      return { content: [{ type: "text", text: JSON.stringify(result.rows, null, 2) }] };
    }

    // ── GET BUDGETS ──
    if (name === "get_budgets") {
      const { month, year } = args;
      let query = `SELECT * FROM "Budget" WHERE "userId" = $1`;
      const params = [CURRENT_USER_ID];
      let i = 2;
      if (month) { query += ` AND month = $${i++}`; params.push(month); }
      if (year)  { query += ` AND year = $${i++}`;  params.push(year); }
      query += ` ORDER BY year DESC, month DESC`;
      const result = await pool.query(query, params);
      return { content: [{ type: "text", text: JSON.stringify(result.rows, null, 2) }] };
    }

    // ── CREATE BUDGET ──
    if (name === "create_budget") {
      const { name: bname, amount, category, month, year } = args;
      const result = await pool.query(
        `INSERT INTO "Budget" (id, name, amount, spent, category, month, year, "createdAt", "updatedAt", "userId")
         VALUES (gen_random_uuid(), $1, $2, 0, $3, $4, $5, NOW(), NOW(), $6)
         RETURNING id`,
        [bname, amount, category, month, year, CURRENT_USER_ID]
      );
      return { content: [{ type: "text", text: `✅ Budget created with ID: ${result.rows[0].id}` }] };
    }

    // ── UPDATE BUDGET ──
    if (name === "update_budget") {
      const { id, name: bname, amount, category } = args;
      const updates = [];
      const params = [];
      let i = 1;
      if (bname)            { updates.push(`name = $${i++}`);     params.push(bname); }
      if (amount !== undefined) { updates.push(`amount = $${i++}`); params.push(amount); }
      if (category)         { updates.push(`category = $${i++}`); params.push(category); }
      updates.push(`"updatedAt" = NOW()`);
      params.push(id, CURRENT_USER_ID);
      await pool.query(
        `UPDATE "Budget" SET ${updates.join(", ")} WHERE id = $${i++} AND "userId" = $${i}`,
        params
      );
      return { content: [{ type: "text", text: `✅ Budget ${id} updated.` }] };
    }

    // ── DELETE BUDGET ──
    if (name === "delete_budget") {
      await pool.query(`DELETE FROM "Budget" WHERE id = $1 AND "userId" = $2`, [args.id, CURRENT_USER_ID]);
      return { content: [{ type: "text", text: `🗑️ Budget ${args.id} deleted.` }] };
    }

    // ── GET LOANS ──
    if (name === "get_loans") {
      const result = await pool.query(
        `SELECT * FROM "Loan" WHERE "userId" = $1 ORDER BY "createdAt" DESC`,
        [CURRENT_USER_ID]
      );
      return { content: [{ type: "text", text: JSON.stringify(result.rows, null, 2) }] };
    }

    // ── ADD LOAN ──
    if (name === "add_loan") {
      const { name: lname, type, principal, outstanding, emi, interest_rate, tenure_months, start_date } = args;
      const result = await pool.query(
        `INSERT INTO "Loan" (id, name, type, principal, outstanding, emi, "interestRate", "tenureMonths", "startDate", "createdAt", "updatedAt", "userId")
         VALUES (gen_random_uuid(), $1, $2, $3, $4, $5, $6, $7, $8, NOW(), NOW(), $9)
         RETURNING id`,
        [lname, type, principal, outstanding, emi, interest_rate, tenure_months, start_date, CURRENT_USER_ID]
      );
      return { content: [{ type: "text", text: `✅ Loan added with ID: ${result.rows[0].id}` }] };
    }

    // ── UPDATE LOAN ──
    if (name === "update_loan") {
      const { id, outstanding, emi, interest_rate } = args;
      const updates = [];
      const params = [];
      let i = 1;
      if (outstanding !== undefined)  { updates.push(`outstanding = $${i++}`);    params.push(outstanding); }
      if (emi !== undefined)          { updates.push(`emi = $${i++}`);            params.push(emi); }
      if (interest_rate !== undefined){ updates.push(`"interestRate" = $${i++}`); params.push(interest_rate); }
      updates.push(`"updatedAt" = NOW()`);
      params.push(id, CURRENT_USER_ID);
      await pool.query(
        `UPDATE "Loan" SET ${updates.join(", ")} WHERE id = $${i++} AND "userId" = $${i}`,
        params
      );
      return { content: [{ type: "text", text: `✅ Loan ${id} updated.` }] };
    }

    // ── DELETE LOAN ──
    if (name === "delete_loan") {
      await pool.query(`DELETE FROM "Loan" WHERE id = $1 AND "userId" = $2`, [args.id, CURRENT_USER_ID]);
      return { content: [{ type: "text", text: `🗑️ Loan ${args.id} deleted.` }] };
    }

  } catch (error) {
    console.error("DB_ERROR:", error);
    return {
      content: [{ type: "text", text: `❌ Database Error: ${error.message}` }],
      isError: true
    };
  }

  throw new Error(`Tool not found: ${name}`);
});

// ─────────────────────────────────────────────
// START SERVER
// ─────────────────────────────────────────────
const transport = new StdioServerTransport();
await server.connect(transport);
console.error("✅ FinFlow MCP Server v2.0 running...");

File: server/mcp_server.py
#!/usr/bin/env python3
"""
FinFlow MCP Server v3.0 — Python / FastMCP 2.x
Connects Claude to the FinFlow PostgreSQL database.

Auth: every request must carry  Authorization: Bearer <token>
      The token is verified by AuthMiddleware (auth.py), which sets
      user_id_var so tool handlers can call uid() to get the caller's ID.

Deploy on Render as a web service; entry-point is main.py (uvicorn).
"""

import os
import json
from typing import Optional

import psycopg2
import psycopg2.extras
from psycopg2 import pool as pg_pool
from fastmcp import FastMCP

from auth import user_id_var  # ContextVar set per-request by AuthMiddleware

# ─────────────────────────────────────────────
# CONFIG
# ─────────────────────────────────────────────

DATABASE_URL = os.environ["DATABASE_URL"]

connection_pool = pg_pool.ThreadedConnectionPool(
    1, 5,
    dsn=DATABASE_URL,
    cursor_factory=psycopg2.extras.RealDictCursor
)

# ─────────────────────────────────────────────
# HELPERS
# ─────────────────────────────────────────────

def uid() -> str:
    """Return the authenticated user_id for the current request."""
    return user_id_var.get()


def get_conn():
    return connection_pool.getconn()


def release_conn(conn):
    connection_pool.putconn(conn)


def query(sql: str, params: tuple = ()):
    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(sql, params)
            conn.commit()
            try:
                rows = cur.fetchall()
                return [dict(r) for r in rows]
            except psycopg2.ProgrammingError:
                return []
    except Exception:
        conn.rollback()
        raise
    finally:
        release_conn(conn)


def rows_to_text(rows: list) -> str:
    return json.dumps(rows, indent=2, default=str)

# ─────────────────────────────────────────────
# FASTMCP APP
# ─────────────────────────────────────────────

mcp = FastMCP(
    name="finflow-backend-agent",
    instructions="FinFlow personal finance assistant. Manages transactions, budgets, and loans."
)

# ═════════════════════════════════════════════
# BALANCE
# ═════════════════════════════════════════════

@mcp.tool(description="Get the current total balance (Income - Expenses)")
def get_balance() -> str:
    rows = query(
        """
        SELECT
            SUM(CASE WHEN type = 'income'  THEN amount ELSE 0 END) AS income,
            SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END) AS expenses
        FROM "Transaction"
        WHERE "userId" = %s
        """,
        (uid(),)
    )
    income   = float(rows[0]["income"]   or 0)
    expenses = float(rows[0]["expenses"] or 0)
    balance  = income - expenses
    return f"Balance: ₹{balance:.2f} (Income: ₹{income:.2f}, Expenses: ₹{expenses:.2f})"

# ═════════════════════════════════════════════
# TRANSACTIONS
# ═════════════════════════════════════════════

@mcp.tool(description="Get transactions with optional filters")
def get_transactions(
    type: Optional[str] = None,
    category: Optional[str] = None,
    from_date: Optional[str] = None,
    to_date: Optional[str] = None,
    limit: int = 50
) -> str:
    sql    = 'SELECT * FROM "Transaction" WHERE "userId" = %s'
    params = [uid()]
    if type:      sql += " AND type = %s";      params.append(type)
    if category:  sql += " AND category = %s";  params.append(category)
    if from_date: sql += " AND date >= %s";     params.append(from_date)
    if to_date:   sql += " AND date <= %s";     params.append(to_date)
    sql += " ORDER BY date DESC LIMIT %s";      params.append(limit)
    rows = query(sql, tuple(params))
    return rows_to_text(rows)


@mcp.tool(description="Log a new income entry")
def add_income(
    amount: float,
    description: str,
    category: str,
    budget_id: Optional[str] = None
) -> str:
    query(
        """
        INSERT INTO "Transaction"
            (id, description, amount, type, category, date, source, "userId", "budgetId")
        VALUES (gen_random_uuid(), %s, %s, 'income', %s, NOW(), 'ai_parsed', %s, %s)
        """,
        (description, amount, category, uid(), budget_id)
    )
    return f'✅ Income of ₹{amount} logged: "{description}" under {category}.'


@mcp.tool(description="Log a new expense entry")
def add_expense(
    amount: float,
    description: str,
    category: str,
    budget_id: Optional[str] = None
) -> str:
    user_id = uid()
    query(
        """
        INSERT INTO "Transaction"
            (id, description, amount, type, category, date, source, "userId", "budgetId")
        VALUES (gen_random_uuid(), %s, %s, 'expense', %s, NOW(), 'ai_parsed', %s, %s)
        """,
        (description, amount, category, user_id, budget_id)
    )
    if budget_id:
        query(
            'UPDATE "Budget" SET spent = spent + %s, "updatedAt" = NOW() WHERE id = %s AND "userId" = %s',
            (amount, budget_id, user_id)
        )
    return f'✅ Expense of ₹{amount} logged: "{description}" under {category}.'


@mcp.tool(description="Update an existing transaction")
def update_transaction(
    id: str,
    amount: Optional[float] = None,
    description: Optional[str] = None,
    category: Optional[str] = None,
    type: Optional[str] = None
) -> str:
    updates, params = [], []
    if amount is not None: updates.append("amount = %s");      params.append(amount)
    if description:        updates.append("description = %s"); params.append(description)
    if category:           updates.append("category = %s");    params.append(category)
    if type:               updates.append("type = %s");        params.append(type)
    if not updates:
        return "No fields to update."
    params.extend([id, uid()])
    query(
        f'UPDATE "Transaction" SET {", ".join(updates)} WHERE id = %s AND "userId" = %s',
        tuple(params)
    )
    return f"✅ Transaction {id} updated."


@mcp.tool(description="Delete a transaction by ID")
def delete_transaction(id: str) -> str:
    query('DELETE FROM "Transaction" WHERE id = %s AND "userId" = %s', (id, uid()))
    return f"🗑️ Transaction {id} deleted."


@mcp.tool(description="Get spending grouped by category for a given month/year")
def get_spending_summary(
    month: Optional[int] = None,
    year: Optional[int] = None
) -> str:
    sql    = 'SELECT category, SUM(amount) AS total FROM "Transaction" WHERE "userId" = %s AND type = \'expense\''
    params = [uid()]
    if month: sql += " AND EXTRACT(MONTH FROM date) = %s"; params.append(month)
    if year:  sql += " AND EXTRACT(YEAR FROM date) = %s";  params.append(year)
    sql += " GROUP BY category ORDER BY total DESC"
    rows = query(sql, tuple(params))
    return rows_to_text(rows)


@mcp.tool(description="Get income vs expense totals month by month")
def get_monthly_trend(months: int = 6) -> str:
    rows = query(
        """
        SELECT
            TO_CHAR(date, 'YYYY-MM') AS month,
            SUM(CASE WHEN type = 'income'  THEN amount ELSE 0 END) AS income,
            SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END) AS expenses
        FROM "Transaction"
        WHERE "userId" = %s
          AND date >= NOW() - (%s || ' months')::INTERVAL
        GROUP BY TO_CHAR(date, 'YYYY-MM')
        ORDER BY month ASC
        """,
        (uid(), str(months))
    )
    return rows_to_text(rows)

# ═════════════════════════════════════════════
# BUDGETS
# ═════════════════════════════════════════════

@mcp.tool(description="Get all budgets, optionally filtered by month/year")
def get_budgets(
    month: Optional[int] = None,
    year: Optional[int] = None
) -> str:
    sql    = 'SELECT * FROM "Budget" WHERE "userId" = %s'
    params = [uid()]
    if month: sql += " AND month = %s"; params.append(month)
    if year:  sql += " AND year = %s";  params.append(year)
    sql += " ORDER BY year DESC, month DESC"
    rows = query(sql, tuple(params))
    return rows_to_text(rows)


@mcp.tool(description="Create a new budget for a category and month")
def create_budget(
    name: str,
    amount: float,
    category: str,
    month: int,
    year: int
) -> str:
    rows = query(
        """
        INSERT INTO "Budget"
            (id, name, amount, spent, category, month, year, "createdAt", "updatedAt", "userId")
        VALUES (gen_random_uuid(), %s, %s, 0, %s, %s, %s, NOW(), NOW(), %s)
        RETURNING id
        """,
        (name, amount, category, month, year, uid())
    )
    return f"✅ Budget created with ID: {rows[0]['id']}"


@mcp.tool(description="Update an existing budget's name, amount, or category")
def update_budget(
    id: str,
    name: Optional[str] = None,
    amount: Optional[float] = None,
    category: Optional[str] = None
) -> str:
    updates, params = [], []
    if name:              updates.append("name = %s");    params.append(name)
    if amount is not None: updates.append("amount = %s"); params.append(amount)
    if category:          updates.append("category = %s"); params.append(category)
    updates.append('"updatedAt" = NOW()')
    params.extend([id, uid()])
    query(
        f'UPDATE "Budget" SET {", ".join(updates)} WHERE id = %s AND "userId" = %s',
        tuple(params)
    )
    return f"✅ Budget {id} updated."


@mcp.tool(description="Delete a budget by ID")
def delete_budget(id: str) -> str:
    query('DELETE FROM "Budget" WHERE id = %s AND "userId" = %s', (id, uid()))
    return f"🗑️ Budget {id} deleted."

# ═════════════════════════════════════════════
# LOANS
# ═════════════════════════════════════════════

@mcp.tool(description="Get all loans for the current user")
def get_loans() -> str:
    rows = query(
        'SELECT * FROM "Loan" WHERE "userId" = %s ORDER BY "createdAt" DESC',
        (uid(),)
    )
    return rows_to_text(rows)


@mcp.tool(description="Add a new loan")
def add_loan(
    name: str,
    type: str,
    principal: float,
    outstanding: float,
    emi: float,
    interest_rate: float,
    tenure_months: int,
    start_date: str
) -> str:
    rows = query(
        """
        INSERT INTO "Loan"
            (id, name, type, principal, outstanding, emi,
             "interestRate", "tenureMonths", "startDate",
             "createdAt", "updatedAt", "userId")
        VALUES (gen_random_uuid(), %s, %s, %s, %s, %s, %s, %s, %s, NOW(), NOW(), %s)
        RETURNING id
        """,
        (name, type, principal, outstanding, emi,
         interest_rate, tenure_months, start_date, uid())
    )
    return f"✅ Loan added with ID: {rows[0]['id']}"


@mcp.tool(description="Update an existing loan (e.g. after EMI payment, update outstanding)")
def update_loan(
    id: str,
    outstanding: Optional[float] = None,
    emi: Optional[float] = None,
    interest_rate: Optional[float] = None
) -> str:
    updates, params = [], []
    if outstanding is not None:   updates.append("outstanding = %s");    params.append(outstanding)
    if emi is not None:           updates.append("emi = %s");            params.append(emi)
    if interest_rate is not None: updates.append('"interestRate" = %s'); params.append(interest_rate)
    updates.append('"updatedAt" = NOW()')
    params.extend([id, uid()])
    query(
        f'UPDATE "Loan" SET {", ".join(updates)} WHERE id = %s AND "userId" = %s',
        tuple(params)
    )
    return f"✅ Loan {id} updated."


@mcp.tool(description="Delete a loan by ID")
def delete_loan(id: str) -> str:
    query('DELETE FROM "Loan" WHERE id = %s AND "userId" = %s', (id, uid()))
    return f"🗑️ Loan {id} deleted."
